include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,519 @@
+   *
+   * Since this class is templated by DATA type, the users DATA class has to
+   * implement the following functions shown below for a quadrature point data
+   * with 2 scalars :
+   * <code>
+   * struct MyQData
+   * {
+   *   // a function which returns the total number of scalars which will be transfered
+   *   static unsigned int number_of_scalars()
+   *   {
+   *     return 2;
+   *   }
+   *
+   *   // a function to pack scalars into a vector
+   *   void pack_scalars(std::vector<double> &scalars) const
+   *   {
+   *     Assert (scalars.size()==2, ExcInternalError());
+   *     scalars[0] = value1;
+   *     scalars[1] = value2;
+   *   }
+   *
+   *   void unpack_scalars(const std::vector<double> &scalars)
+   *   {
+   *     Assert (scalars.size() ==2, ExcInternalError());
+   *     value1 = scalars[0];
+   *     value2 = scalars[1];
+   *   }
+   * };
+   * </code>
+   * This approach can be extended to quadrature point data with Tensors of
+   * arbitrary order, although with a little bit more work in packing and
+   * unpacking of data.
+   *
+   * @note During mesh balancing children of a single cell can be attributed
+   * to different processors. As a corollary, their parent cells can exist on
+   * multiple processors.
+   *
+   *
+   * @note The functionality provided by this class can alternatively be achieved
+   * with the parallel::distributed::SolutionTransfer . However, that would require
+   * steps: (i) creating an auxiliary DoFHandler with a (discontinuous Galerkin)
+   * FiniteElement which has enough components to represent all data stored at the
+   * quadrature points data. (ii) projecting the data to the FiniteElement space
+   * and thereby storing results in vectors; (iii) using the parallel::distributed::SolutionTransfer
+   * to project those vectors to the new mesh; (iv) and finally projecting the
+   * data back to the quadrature points on the new mesh via FEValues class.
+   * The QPManager class aims at simplifying the whole process only by requiring
+   * that the quadrature point data class is derived from the QuadraturePointData
+   * and thereby implements functions for packing and unpacking its data into a
+   * vector. Everything else will be done automatically.
+   *
+   * @author Denis Davydov, Jean-Paul Pelteret, 2016
+   */
+  template<int dim, typename DATA>
+  class QPManager : public QPManagerBase<typename parallel::distributed::Triangulation<dim>::cell_iterator, DATA>
+  {
+  public:
+    /**
+     * A typedef for a cell.
+     */
+    typedef typename parallel::distributed::Triangulation<dim>::cell_iterator CELL;
+
+    /**
+     * Constructor which takes the FiniteElement @p projection_fe , the quadrature
+     * rule @p lhs_quadrature used to integrate its local mass matrix and
+     * finally the quadrature rule @p rhs_quadrature used to store DATA .
+     */
+    QPManager(const FiniteElement<dim> &projection_fe,
```

Comment:
Note: Here its required that p-refinement doesn't occur.

For h-p refinement (with `hp::DoFHandler`),  only one FECollection and one QCollection need be sent in for the construction of the mass matrix. One can use `active_fe_index()` to fetch the relevant base quadrature rules for the RHS.